internal/storage/streamed_chunk_uploader.go

```diff
@@ -0,0 +1,206 @@
+	if obj == nil || req == nil {
+		return nil, fmt.Errorf("nil obj or req")
+	}
+
+	if obj.ObjectName() != req.Name {
+		return nil, fmt.Errorf("names of passed object-handle (%s) and CreateObjectRequest (%s) don't match", obj.ObjectName(), req.Name)
+	}
+
+	if req.GenerationPrecondition != nil && *req.GenerationPrecondition != 0 {
+		return nil, fmt.Errorf("request received for pre-existing object %s, supported only for new objects", req.Name)
+	}
+
+	if writeChunkSize <= 0 {
+		return nil, fmt.Errorf("writeChunkSize <= 0 not supported")
+	}
+
+	// raw initialization
+	sow := StreamedChunkUploader{}
+	sow.state = Uninitialized
+
+	// store references to necessary parameters
+	sow.objectName = obj.ObjectName()
+	sow.userProgressFunc = progressFunc
+
+	// Create a NewWriter with the requested attributes, using Go Storage Client.
+	// Chuck size for resumable upload is default i.e. 16MB.
+	wc := obj.NewWriter(ctx)
+	// NewWriter never returns nil, so no nil-check here.
+
+	wc = storageutil.SetAttrsInWriter(wc, req)
+	wc.ChunkSize = writeChunkSize
+	wc.ProgressFunc = func(n int64) {
+		sow.totalWriteSucceededSoFar = n
+		logger.Debugf("%d bytes copied so far for object/file %s. chunk-size = %d", n, req.Name, wc.ChunkSize)
+
+		if sow.userProgressFunc != nil {
+			sow.userProgressFunc(n)
+		}
+	}
+	sow.wc = wc
+	sow.state = Initialized
+
+	return &sow, nil
+}
+
+func (sow StreamedChunkUploader) ChunkSize() (int, error) {
+	if nil == sow.wc {
+		return 0, fmt.Errorf("writer not initialized")
+	}
+	return sow.wc.ChunkSize, nil
+}
+
+func (sow StreamedChunkUploader) State() UploaderState {
+	return sow.state
+}
+
+func (sow StreamedChunkUploader) BytesUploadInitiatedSoFar() int64 {
+	return sow.totalWriteInitiatedSoFar
+}
+
+func (sow StreamedChunkUploader) BytesUploadedSoFar() int64 {
+	return sow.totalWriteSucceededSoFar
+}
+
+func (sow *StreamedChunkUploader) readyToWrite() bool {
+	switch sow.state {
+	case Initialized:
+		fallthrough
+	case Waiting:
+		fallthrough
+	case Writing:
+		return true
+	default:
+		return false
+	}
+}
+
+func (sow *StreamedChunkUploader) readyToClose() bool {
+	switch sow.state {
+	case Initialized:
+		fallthrough
+	case Waiting:
+		fallthrough
+	case Writing:
+		fallthrough
+	case WriteError:
+		return true
+	default:
+		return false
+	}
+}
+
+// Write uploads the given content
+// to underlying GCS object with the initial configures chunk-size,
+// number of bytes, an object writer, and a content-reader.
+// It is assumed that all the relevant attributes of the object
+// writer have been already by a previous calls to CreateObject
+// which created the object-writer.
+func (sow *StreamedChunkUploader) UploadChunkAsync(contents io.Reader) error {
+	if !sow.readyToWrite() || (sow.wc == nil) {
+		return fmt.Errorf("writer not ready to write: object: %s, status = %v, writer: %v", sow.objectName, sow.state, sow.wc)
+	}
+
+	n, err := io.Copy(sow.wc, contents)
+	sow.totalWriteInitiatedSoFar += n
+	if n == 0 || err == io.EOF {
```

Comment:
why are you converting n==0 to io.EOF? just make it a nop when no data is passed?

Reply:
Done. Removed the `n==0` case.

Reply:
Yeah, I was wrong earlier, io.EOF is never returned from io.Copy in successful case unless size is specified. Since we're specifying size here, we can't get io.EOF. So, I've removed any special handling for io.EOF.

As for n==0, I've treated that a NOP now.